drivers/wdm/audio/backpln/portcls/pin_wavepci.cpp

```diff
@@ -588,22 +646,33 @@ CPortPinWavePci::DeviceIoControl(
 
     IoStack = IoGetCurrentIrpStackLocation(Irp);
 
-    if (IoStack->Parameters.DeviceIoControl.IoControlCode == IOCTL_KS_PROPERTY)
+    switch (IoStack->Parameters.DeviceIoControl.IoControlCode)
     {
-        return HandleKsProperty(Irp);
-    }
-    else if (IoStack->Parameters.DeviceIoControl.IoControlCode == IOCTL_KS_WRITE_STREAM || IoStack->Parameters.DeviceIoControl.IoControlCode == IOCTL_KS_READ_STREAM)
-    {
-       return HandleKsStream(Irp);
-    }
+        case IOCTL_KS_PROPERTY:
+            return HandleKsProperty(Irp);
 
-    UNIMPLEMENTED;
+        case IOCTL_KS_ENABLE_EVENT:
+            return PcHandleEnableEventWithTable(Irp, &m_Descriptor);
+
+        case IOCTL_KS_DISABLE_EVENT:
+            return PcHandleDisableEventWithTable(Irp, &m_Descriptor);
+
+        case IOCTL_KS_RESET_STATE:
+            return HandleKsReset(Irp);
+
+        case IOCTL_KS_READ_STREAM:
+        case IOCTL_KS_WRITE_STREAM:
+            return HandleKsStream(Irp);
+
+        default:
+            return KsDefaultDeviceIoCompletion(DeviceObject, Irp);
```

Comment:
Why using a function from ksfilter driver?

Reply:
Because I didn't find an appropriate equivalent in portcls itself or in kernel. Is there any? :thinking:

Reply:
You already handle most of the cases from `KsDefaultDeviceIoCompletion` here. Seems like you'd just need a case for `IOCTL_KS_METHOD` that completes the IRP with `STATUS_PROPSET_NOT_FOUND`, and otherwise complete it with `STATUS_INVALID_DEVICE_REQUEST`.